codespan-reporting/examples/custom_files.rs

```diff
@@ -0,0 +1,104 @@
+//! This example shows how to make a custom file database for use with codespan.
+//!
+//! This implements a custom database that uses `Arc<str>`s for source storage,
+//! and 32-bit file-ids.
+
+use codespan_reporting::files;
+use std::ops::Range;
+use std::sync::Arc;
+
+fn main() {
+    todo!()
+}
+
+/// A file that is backed by an `Arc<String>`.
+#[derive(Debug, Clone)]
+struct ArcFile {
+    /// The name of the file.
+    name: String,
+    /// The source code of the file.
+    source: Arc<str>,
```

Comment:
Perhaps this is a bit much of a jump in complexity? Perhaps we should just use a `String` here?